### PR TITLE
drivers/timers: Add weak_function for up_ function

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
so the user could replace the implementation if need

## Impact
Can replace the default implementation

## Testing
Pass CI

### CODE REVIEW

**Comment by anchao** [comment]
LGTM

File: drivers/timers/arch_alarm.c [hunk 1]
```diff
@@ -204,7 +204,7 @@ void up_alarm_set_lowerhalf(FAR struct oneshot_lowerhalf_s *lower)
  ****************************************************************************/
 
 #ifdef CONFIG_CLOCK_TIMEKEEPING
-int up_timer_getcounter(FAR uint64_t *cycles)
+int weak_function up_timer_getcounter(FAR uint64_t *cycles)
```
Comment (pkarashchenko) [inline drivers/timers/arch_alarm.c:207]: If `up_timer_getcounter` becomes `weak`, then shouldn't we change
```
ret = up_timer_getcounter(&counter);
```
to
```
if (up_timer_getcounter != NULL)
  {
    ret = up_timer_getcounter(&counter);
  }
else
  {
    ret = ERROR;
  }
```
???

Comment (xiaoxiang781216) [reply drivers/timers/arch_alarm.c:207]: The intention of weak is to allow the board code overwrite the default implementation, but not to make it optional.

Comment (pkarashchenko) [reply drivers/timers/arch_alarm.c:207]: Thank you for explanation.

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Pull request merged**